### PR TITLE
[Fix #8027] Avoid avoidable `File.directory?` in RuboCop::ConfigStore

### LINKED ISSUE #8027: Performance: avoid avoidable `File.directory?`
We spend 3% of the time calling `File.directory?` when we know it's a file and not a directory...

```sh
$ stackprof tmp/stackprof-cpu-*.dump --method 'RuboCop::ConfigStore#for'
RuboCop::ConfigStore#for (/Users/mal/rubocop/lib/rubocop/config_store.rb:32)
  samples:   860 self (2.7%)  /    865 total (2.7%)
  callers:
     765  (   88.4%)  RuboCop::Runner#inspect_file
      52  (    6.0%)  RuboCop::ResultCache#file_checksum
      20  (    2.3%)  RuboCop::Runner#file_offense_cache
      13  (    1.5%)  RuboCop::Runner#get_processed_source
      10  (    1.2%)  RuboCop::ResultCache.cache_root
       5  (    0.6%)  RuboCop::ResultCache.allow_symlinks_in_cache_location?
  callees (5 total):
       5  (  100.0%)  RuboCop::ConfigLoader.configuration_file_for
  code:
                                  |    32  |     def for(file_or_dir)
    1    (0.0%) /     1   (0.0%)  |    33  |       return @options_config if @options_config
                                  |    34  | 
  859    (2.7%) /   859   (2.7%)  |    35  |       dir = if File.directory?(file_or_dir)
                                  |    36  |               file_or_dir
                                  |    37  |             else
                                  |    38  |               File.dirname(file_or_dir)
                                  |    39  |             end
    5    (0.0%)                   |    40  |       @path_cache[dir] ||= ConfigLoader.configuration_file_for(dir)
                                  |    41  |       path = @path_cache[dir]
```

### PR DESCRIPTION
**Pull request by tejasbubane**
Closes #8027 

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [ ] Added an entry to the [Changelog](https://github.com/rubocop-hq/rubocop/blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop-hq/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Run `bundle exec rake default`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Comment by tejasbubane** [comment]
Not sure if performance improvements need separate changelog entries.

File: lib/rubocop/result_cache.rb [hunk 1]
```diff
@@ -33,7 +33,7 @@ class << self
 
       def requires_file_removal?(file_count, config_store)
         file_count > 1 &&
-          file_count > config_store.for('.').for_all_cops['MaxFilesInCache']
+          file_count > config_store.for_file('.').for_all_cops['MaxFilesInCache']
```
Comment (marcandre) [inline lib/rubocop/result_cache.rb:36]: Should be `for_dir`, no?

File: lib/rubocop/result_cache.rb [hunk 2]
```diff
@@ -60,7 +60,7 @@ def remove_files(files, dirs, remove_count)
     end
 
     def self.cache_root(config_store)
-      root = config_store.for('.').for_all_cops['CacheRootDirectory']
+      root = config_store.for_file('.').for_all_cops['CacheRootDirectory']
```
Comment (marcandre) [inline lib/rubocop/result_cache.rb:63]: Same?

File: lib/rubocop/result_cache.rb [hunk 3]
```diff
@@ -72,7 +72,7 @@ def self.cache_root(config_store)
     end
 
     def self.allow_symlinks_in_cache_location?(config_store)
-      config_store.for('.').for_all_cops['AllowSymlinksInCacheRootDirectory']
+      config_store.for_file('.').for_all_cops['AllowSymlinksInCacheRootDirectory']
```
Comment (marcandre) [inline lib/rubocop/result_cache.rb:75]: Again

**Comment by marcandre** [comment]
Looks good, apart from the cases I'm pointing out.

Changelog: depends how major the optimization is. In this case though, it's more that we are adding `for_dir` and `for_file`, that should be in the ChangeLog...

Maybe you want to add a comment to `for` suggesting to use `for_dir`/`for_file` if possible.

**Comment by tejasbubane** [comment]
@marcandre Fixed.

File: lib/rubocop/config_store.rb [hunk 4]
```diff
@@ -29,14 +29,24 @@ def force_default_config!
...
     end
 
-    def for(file_or_dir)
-      return @options_config if @options_config
+    def for_file(file)
+      for_dir(File.dirname(file))
+    end
 
+    # If type (file/dir) is known beforehand,
+    # prefer using #for_fir or #for_dir for improved performance
```
Comment (marcandre) [inline lib/rubocop/config_store.rb:37]: ^for_fir^for_file

Comment (tejasbubane) [reply lib/rubocop/config_store.rb]: My bad. Fixed.

**Comment by marcandre** [comment]
Super, thanks!

**Pull request merged**